### PR TITLE
feat: support default value when inserting data

### PR DESCRIPTION
**Pull request by lizhemingi**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

support default value when inserting data.

Please explain IN DETAIL what the changes are in this PR and why they are needed:

- Change `Identifier` without quoted to `Placeholder` instead of `SingleQuotedString` when inserting data.
- Replace `Placeholder` with `default` and `DEFAULT` value to the default value of corresponding column if possible.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.
image

## Refer to a related PR or issue link (optional)
#759

### CODE REVIEW

**Comment by lizhemingi** [comment]
@killme2008 PTAL. 

Sorry for the slow progress, since I was suffering from covid-19 and busy work last few weeks.

**Comment by killme2008** [comment]
> @killme2008 PTAL.
> 
> Sorry for the slow progress, since I was suffering from covid-19 and busy work last few weeks.

Wish you health. We'll review it ASAP. Thanks a lot.

File: src/datanode/src/error.rs [hunk 1]
```diff
@@ -283,6 +284,34 @@ pub enum Error {
...
+        column: String,
+        #[snafu(backtrace)]
+        source: datatypes::error::Error,
+    },
+
+    #[snafu(display(
+        "No valid default value can be build automatically, column: {}",
+        column,
+    ))]
+    ColumnNoneDefaultValue { column: String },
```
Comment (killme2008) [inline src/datanode/src/error.rs:314]: Please add `backtrace: Backtrace` field.

File: src/datanode/src/sql/insert.rs [hunk 2]
```diff
@@ -125,21 +117,39 @@ impl SqlHandler {
...
     builder: &mut Box<dyn MutableVector>,
 ) -> Result<()> {
-    let value = statements::sql_value_to_value(column_name, data_type, sql_val)
-        .context(ParseSqlValueSnafu)?;
-    builder.push_value_ref(value.as_value_ref()).unwrap();
+    let value = if replace_default(sql_val) {
+        column_schema
+            .create_default()
+            .context(ColumnDefaultValueSnafu {
+                column: column_schema.name.to_owned(),
```
Comment (killme2008) [inline src/datanode/src/sql/insert.rs:135]: We prefer using `to_string` to replace `to_ownded`, it's more explicit.

File: src/datanode/src/sql/insert.rs [hunk 3]
```diff
@@ -125,21 +117,39 @@ impl SqlHandler {
...
+            })?
+            .context(ColumnNoneDefaultValueSnafu {
+                column: column_schema.name.to_owned(),
```
Comment (killme2008) [inline src/datanode/src/sql/insert.rs:138]: The same with above.

File: src/datanode/src/sql/insert.rs [hunk 4]
```diff
@@ -125,21 +117,39 @@ impl SqlHandler {
...
+
+fn add_value_to_vector(value: Value, builder: &mut Box<dyn MutableVector>) -> Result<()> {
+    builder
+        .push_value_ref(value.as_value_ref())
+        .context(BuildVectorSnafu { value })
+}
 
-    Ok(())
+fn replace_default(sql_val: &SqlValue) -> bool {
+    matches!(sql_val, SqlValue::Placeholder(s) if s == "default")
```
Comment (killme2008) [inline src/datanode/src/sql/insert.rs:154]: Make `"deault"` to be a constant variable.

File: src/frontend/src/error.rs [hunk 5]
```diff
@@ -369,6 +369,23 @@ pub enum Error {
...
+        column: String,
+        #[snafu(backtrace)]
+        source: datatypes::error::Error,
+    },
+
+    #[snafu(display(
+        "No valid default value can be build automatically, column: {}",
+        column,
+    ))]
+    ColumnNoneDefaultValue { column: String },
```
Comment (killme2008) [inline src/frontend/src/error.rs:388]: Adds `backtrace` too.

File: src/frontend/src/sql.rs [hunk 6]
```diff
@@ -91,21 +83,38 @@ pub(crate) fn insert_to_request(table: &TableRef, stmt: Insert) -> Result<Insert
...
 fn add_row_to_vector(
-    column_name: &str,
-    data_type: &ConcreteDataType,
+    column_schema: &ColumnSchema,
     sql_val: &SqlValue,
     builder: &mut Box<dyn MutableVector>,
 ) -> Result<()> {
-    let value = statements::sql_value_to_value(column_name, data_type, sql_val)
-        .context(error::ParseSqlSnafu)?;
+    let value = if replace_default(sql_val) {
```
Comment (killme2008) [inline src/frontend/src/sql.rs:96]: Looks like these logics are almost the same with `insert.rs` in datanode, we can refactor them later. can you add a TODO here? Thanks.

File: src/sql/src/error.rs [hunk 7]
```diff
@@ -124,6 +124,9 @@ pub enum Error {
         #[snafu(backtrace)]
         source: api::error::Error,
     },
+
+    #[snafu(display("Invalid sql value: {}", value))]
+    InvalidSqlValue { value: String },
```
Comment (killme2008) [inline src/sql/src/error.rs:129]: Adds `backtrace`.

File: src/sql/src/statements.rs [hunk 8]
```diff
@@ -222,6 +223,7 @@ pub fn sql_value_to_value(
             parse_string_to_value(column_name, s.to_owned(), data_type)?
         }
         SqlValue::HexStringLiteral(s) => parse_hex_string(s)?,
+        SqlValue::Placeholder(s) => return InvalidSqlValueSnafu { value: s }.fail(),
```
Comment (killme2008) [inline src/sql/src/statements.rs:226]: `return` is unnecessary.

File: src/sql/src/statements/insert.rs [hunk 9]
```diff
@@ -146,4 +146,22 @@ mod tests {
             _ => unreachable!(),
         }
     }
+
+    #[test]
+    fn test_insert_value_with_default() {
+        use crate::statements::statement::Statement;
+
+        // insert "default"
+        let sql = "INSERT INTO my_table VALUES(default)";
```
Comment (killme2008) [inline src/sql/src/statements/insert.rs:155]: If the `default` is in uppercase, the parsed result is `default` in lowercase too? Maybe we should test it.

File: src/frontend/src/sql.rs (hunk 6 above)
Comment (lizhemingi) [reply src/frontend/src/sql.rs:96]: There is already a TODO in `frontend/src/sql.rs`. 

// TODO(fys): Extract the common logic in datanode and frontend in the future.

And maybe I can take that too after is PR.

File: src/sql/src/statements.rs (hunk 8 above)
Comment (lizhemingi) [reply src/sql/src/statements.rs:226]: Seems not. The whole match is wrapped in an `Ok` as return value.

File: src/sql/src/statements/insert.rs (hunk 9 above)
Comment (lizhemingi) [reply src/sql/src/statements/insert.rs:155]: Had a simple test, the parsed result stays `DEFAULT`, will update the code and unit test later.

Nice catch.

File: src/datanode/src/sql/insert.rs [hunk 10]
```diff
@@ -125,21 +117,39 @@ impl SqlHandler {
...
+        statements::sql_value_to_value(&column_schema.name, &column_schema.data_type, sql_val)
+            .context(ParseSqlSnafu)?
+    };
+    add_value_to_vector(value, builder)
+}
+
+fn add_value_to_vector(value: Value, builder: &mut Box<dyn MutableVector>) -> Result<()> {
+    builder
+        .push_value_ref(value.as_value_ref())
+        .context(BuildVectorSnafu { value })
```
[the remark this answers is not in the document]
Comment (lizhemingi) [reply src/datanode/src/sql/insert.rs]: Also change the way in `src/frontend/src/sql.rs`

**Comment by lizhemingi** [comment]
@killme2008 @killme2008 Please take another look.

**Review by killme2008 — APPROVED** [review]
LGTM

**Review by waynexia — APPROVED** [review]
Looks good to me, thanks @lizhemingi. I'll add some cases to sqlness test to cover this change.

**Pull request merged**